File: src/ir/cursorIr.ts
import type { TerminalRecord } from "../enrich/toolMetrics.js";
import type { CursorEvent, RawTrajectory, Substep, TrajectoryIR, TrajectoryStep, ToolExecutionMetrics } from "../types/index.js";
import { validateIr } from "./schema.js";
import { applySessionWallMetrics, buildCursorSessionWallMetrics } from "./sessionMetrics.js";
import { buildStepTelemetry, extractCursorStepFields } from "./stepTelemetry.js";

const USER_QUERY_RE = /<user_query>\s*(.*?)\s*<\/user_query>/s;

export interface CursorIrOptions {
  terminals?: TerminalRecord[];
}

function extractUserText(content: string): string {
  const m = USER_QUERY_RE.exec(content);
  return (m ? m[1] : content).trim();
}

function toolRole(name: string, inp: Record<string, unknown>): string {
  if (name === "CallMcpTool") {
    const server = String(inp.server ?? inp.mcpServer ?? "unknown");
    const tool = String(inp.toolName ?? "");
    return `mcp:${server}${tool ? `/${tool}` : ""}`;
  }
  return `tool:${name}`;
}

function toolContent(inp: Record<string, unknown>): string {
  try {
    return JSON.stringify(inp).slice(0, 4000);
  } catch {
    return String(inp).slice(0, 4000);
  }
}

function parseAssistantContent(
  contentList: NonNullable<NonNullable<CursorEvent["message"]>["content"]>,
  stepIndex: number,
  metricsMap?: Map<string, ToolExecutionMetrics>
): Substep[] {
  const substeps: Substep[] = [];
  let subIdx = 0;
  for (const item of contentList) {
    if (item.type === "text") {
      const text = String(item.text ?? "").trim();
      if (text) {
        subIdx++;
        substeps.push({ sub_index: subIdx, role: "assistant", content: text.slice(0, 8000) });
      }
    } else if (item.type === "tool_use") {
      const name = String(item.name ?? "unknown");
      const inp = (item.input ?? {}) as Record<string, unknown>;
      subIdx++;
      substeps.push({
        sub_index: subIdx,
        role: toolRole(name, inp),
        content: toolContent(inp),
        tool_name: name,
        tool_input: inp,
        execution: metricsMap?.get(`${stepIndex}:${subIdx}`),
      });
    }
  }
  return substeps;
}

export function cursorIr(
  trajectories: RawTrajectory[],
  metricsMap?: Map<string, ToolExecutionMetrics>,
  toolEfficiency?: Record<string, unknown>,
  options: CursorIrOptions = {},
): TrajectoryIR[] {
  const out: TrajectoryIR[] = [];

  for (const traj of trajectories) {
    const events = traj.events as CursorEvent[];
    let instruction = traj.instruction ?? "";
    let userTurn = 0;
    const steps: TrajectoryStep[] = [];
    let stepIdx = 0;

    for (const event of events) {
      const role = event.role;
      const contentList = event.message?.content ?? [];

      if (role === "user") {
        userTurn++;
        for (const item of contentList) {
          if (item.type === "text") {
            const text = extractUserText(String(item.text ?? ""));
            if (text && !instruction) instruction = text.slice(0, 2000);
          }
        }
        continue;
      }

      if (role !== "assistant") continue;

      const substeps = parseAssistantContent(contentList, stepIdx, metricsMap);
      if (!substeps.length) continue;

      stepIdx++;
      steps.push({
        index: stepIdx,
        telemetry: buildStepTelemetry(substeps, userTurn, extractCursorStepFields),
        substeps,
      });
    }

    const wall = buildCursorSessionWallMetrics(traj._source_path, events, options.terminals ?? []);

    const ir: TrajectoryIR = {
      trajectory_id: traj.trajectory_id ?? "unknown",
      source: "cursor",
      instruction,
      metadata: applySessionWallMetrics({
        source_path: traj._source_path,
        event_count: events.length,
        step_count: steps.length,
        user_turns: userTurn,
        tool_efficiency: toolEfficiency,
      }, wall),
      steps,
    };
    validateIr(ir);
    out.push(ir);
  }

  return out;
}

File: doctor/ir/cursor_ir.py
"""Convert Cursor JSONL events to canonical Trajectory IR."""

from __future__ import annotations

import json
import re
from typing import Any

from doctor.ir.schema import validate_ir

USER_QUERY_RE = re.compile(r"<user_query>\s*(.*?)\s*</user_query>", re.DOTALL)


def _extract_user_text(content: str) -> str:
    m = USER_QUERY_RE.search(content)
    if m:
        return m.group(1).strip()
    return content.strip()


def _tool_role(name: str, inp: dict[str, Any]) -> str:
    if name == "CallMcpTool":
        server = inp.get("server") or inp.get("mcpServer") or "unknown"
        tool = inp.get("toolName") or ""
        return f"mcp:{server}" + (f"/{tool}" if tool else "")
    return f"tool:{name}"


def _tool_content(name: str, inp: dict[str, Any]) -> str:
    try:
        return json.dumps(inp, ensure_ascii=False)[:4000]
    except Exception:
        return str(inp)[:4000]


def _parse_assistant_content(content_list: list[dict[str, Any]]) -> list[dict[str, Any]]:
    substeps: list[dict[str, Any]] = []
    sub_idx = 0
    for item in content_list:
        t = item.get("type")
        if t == "text":
            text = str(item.get("text") or "").strip()
            if text:
                sub_idx += 1
                substeps.append({
                    "sub_index": sub_idx,
                    "role": "assistant",
                    "content": text[:8000],
                })
        elif t == "tool_use":
            name = str(item.get("name") or "unknown")
            inp = item.get("input") or {}
            sub_idx += 1
            substeps.append({
                "sub_index": sub_idx,
                "role": _tool_role(name, inp),
                "content": _tool_content(name, inp),
                "tool_name": name,
                "tool_input": inp,
            })
    return substeps


def _step_telemetry(substeps: list[dict[str, Any]], user_turn: int) -> dict[str, Any]:
    tool_names: list[str] = []
    mcp_servers: list[str] = []
    shell_cmds: list[str] = []
    grep_patterns: list[str] = []
    read_paths: list[str] = []
    skill_reads: list[str] = []

    for sub in substeps:
        role = sub.get("role", "")
        inp = sub.get("tool_input") or {}
        name = sub.get("tool_name") or ""

        if role.startswith("tool:"):
            tool_names.append(name or role.split(":", 1)[-1])
        if role.startswith("mcp:"):
            mcp_servers.append(role.split(":", 1)[-1].split("/")[0])

        if name == "Shell":
            cmd = str(inp.get("command") or "")
            if cmd:
                shell_cmds.append(cmd.strip())
        elif name == "Grep":
            pat = str(inp.get("pattern") or "")
            if pat:
                grep_patterns.append(pat)
        elif name == "Read":
            path = str(inp.get("path") or "")
            if path:
                read_paths.append(path)
                if "SKILL.md" in path or "/skills/" in path:
                    skill_reads.append(path)

    return {
        "user_turn": user_turn,
        "tool_count": len([s for s in substeps if s.get("role", "").startswith(("tool:", "mcp:"))]),
        "mcp_count": len(mcp_servers),
        "shell_count": len(shell_cmds),
        "read_count": len(read_paths),
        "grep_count": len(grep_patterns),
        "assistant_chars": sum(len(s.get("content", "")) for s in substeps if s.get("role") == "assistant"),
        "tool_names": tool_names,
        "mcp_servers": mcp_servers,
        "shell_cmds": shell_cmds,
        "grep_patterns": grep_patterns,
        "read_paths": read_paths,
        "skill_reads": skill_reads,
    }


def cursor_ir(trajectories: list[dict[str, Any]]) -> list[dict[str, Any]]:
    out: list[dict[str, Any]] = []

    for traj in trajectories:
        trajectory_id = str(traj.get("trajectory_id") or "unknown")
        events = traj.get("events") or []
        source_path = traj.get("_source_path", "")

        instruction = str(traj.get("instruction") or "")
        user_turn = 0
        steps: list[dict[str, Any]] = []
        step_idx = 0

        for event in events:
            role = event.get("role")
            msg = event.get("message") or {}
            content_list = msg.get("content") or []

            if role == "user":
                user_turn += 1
                for item in content_list:
                    if item.get("type") == "text":
                        text = _extract_user_text(str(item.get("text") or ""))
                        if text and not instruction:
                            instruction = text[:2000]
                continue

            if role != "assistant":
                continue

            substeps = _parse_assistant_content(content_list)
            if not substeps:
                continue

            step_idx += 1
            telemetry = _step_telemetry(substeps, user_turn)
            steps.append({
                "index": step_idx,
                "telemetry": telemetry,
                "substeps": substeps,
            })

        ir = {
            "trajectory_id": trajectory_id,
            "source": "cursor",
            "instruction": instruction,
            "metadata": {
                "source_path": source_path,
                "event_count": len(events),
                "step_count": len(steps),
                "user_turns": user_turn,
            },
            "steps": steps,
        }
        validate_ir(ir)
        out.append(ir)

    return out
